pkg/ebpf/map.go

```diff
@@ -70,13 +71,42 @@ func (m *Map) OpenOrCreate() error {
 	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
 	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
 
+	path := bpf.MapPath(m.spec.Name)
+
 	newMap, err := ciliumebpf.NewMapWithOptions(m.spec, opts)
 	if err != nil {
-		return fmt.Errorf("unable to create map: %w", err)
+		if !errors.Is(err, ciliumebpf.ErrMapIncompatible) {
+			return fmt.Errorf("unable to create map: %w", err)
+		}
+
+		// There already exists a pinned map but it has a different
+		// configuration (e.g different type, k/v size or flags).
+		// Try to delete and recreate it.
+
+		log.WithField("map", m.spec.Name).
+			Warn("cannot load pinned map as it's incompatible with the spec provided. Deleting and recreating it.")
+
+		oldMap, err := ciliumebpf.LoadPinnedMap(path, &opts.LoadPinOptions)
+		if err != nil {
+			return fmt.Errorf("cannot load pinned map %s: %w", m.spec.Name, err)
+		}
+
+		if err = oldMap.Unpin(); err != nil {
+			return fmt.Errorf("cannot unpin map %s: %w", m.spec.Name, err)
+		}
+
+		if err = oldMap.Close(); err != nil {
+			return fmt.Errorf("cannot close map %s: %w", m.spec.Name, err)
+		}
```

Comment:
The error handling seems a bit unusual here: As soon as `LoadPinnedMap()` has returned without error, the fd is open. If the `Unpin()` theoretically failed then we wouldn't close the map. Should we just use `defer` here?

Also not sure how `oldMap.Close()` can fail but if we were able to `Unpin()` the old one, do we really need to abort the creation of the new map in that error condition? Maybe we can just log a warning and continue? (idea being if there is some really weird behaviour here then at least it will show up in the logs, but it shouldn't prevent the broader `OpenOrCreate()` operation that's being requested)